lib/smooth/core/io/i2c/I2CMasterDevice.cpp

```diff
@@ -24,77 +24,148 @@ limitations under the License.
-                               core::util::FixedBufferBase<uint8_t>& data,
+    // Write address and 8 bit slave register followed by write read address and read data
+    // NOTE: This function does not support slave clock-stretching
+    bool I2CMasterDevice::read(uint8_t address,
+                               uint8_t slave_register,
+                               FixedBufferBase<uint8_t>& data,
                                bool use_restart_signal,
                                bool end_with_nack)
     {
+        // Create a threadsafe command link
+        I2CCommandLink link(*this);
+
+        // Set the process name that will be used by log_error
+        set_process_name(read_tag);
+
+        // Convert slave register to FixedBuffer
+        FixedBuffer<uint8_t, 1> slave_reg_addr;
+        slave_reg_addr[0] = slave_register;
+
+        return write_followed_by_read(link, address, slave_reg_addr, data, use_restart_signal, end_with_nack, 0);
+    }
+
+    // Write address and 8 bit slave register followed by write read address and read data.
+    // NOTE: This function supports slave clock-stretching
+    bool I2CMasterDevice::read8(uint8_t address,
+                                uint8_t slave_register,
+                                FixedBufferBase<uint8_t>& data,
+                                bool use_restart_signal,
+                                bool end_with_nack,
+                                int scl_timeout)
+    {
+        // Create a threadsafe command link
+        I2CCommandLink link(*this);
+
+        // Set the process name that will be used by log_error
+        set_process_name(read8_tag);
+
+        // Convert slave register to FixedBuffer
+        FixedBuffer<uint8_t, 1> slave_reg;
+        slave_reg[0] = slave_register;
+
+        return write_followed_by_read(link, address, slave_reg, data, use_restart_signal, end_with_nack, scl_timeout);
+    }
+
+    // Write address and 16 bit slave register followed by write read address and read data.
+    // NOTE: This function supports slave clock-stretching
+    bool I2CMasterDevice::read16(uint8_t address,
+                                 uint16_t slave_register,
+                                 FixedBufferBase<uint8_t>& data,
+                                 bool use_restart_signal,
+                                 bool end_with_nack,
+                                 int scl_timeout)
+    {
+        // Create a threadsafe command link
         I2CCommandLink link(*this);
 
+        // Set the process name that will be used by log_error
+        set_process_name(read16_tag);
+
+        // Convert slave register to FixedBuffer
+        FixedBuffer<uint8_t, 2> slave_reg;
+        slave_reg[0] = static_cast<uint8_t>(slave_register >> 8);
+        slave_reg[1] = static_cast<uint8_t>(slave_register & 0x00FF);
+
+        return write_followed_by_read(link, address, slave_reg, data, use_restart_signal, end_with_nack, scl_timeout);
```

Comment:
As already noted, |= is not the correct operator to use for the chained calls throughout this function. If the first call fails, we shouldn't continue the operation.